STANFORD-419 | CTA cards block

yvonnetangsu opened the pull request:
# READY FOR REVIEW

# Summary
- CTA Cards Block component

# Review By (Date)
- Whenever you need it

# Criticality
- 5

# Review Tasks

## Setup tasks and/or behavior to test

1. See my test page
https://sug-web.matrix.squiz.cloud/report/development/yvonnes-test-page
2. Test different responsive breakpoints
3. Test both light and dark mode
4. Change background to #f4f4f4 (fog light) in the inspector and see that the card still look good - this is simulating the future Campaign page. Below is my test:
![Screenshot 2024-03-25 at 4 31 02 PM](https://github.com/SU-UComm/stanford-report/assets/42749717/956fbe6e-8656-4975-b99d-86ac4c15da14)

5. Add the component on your own test page and see the user experience
6. Look at code

# Associated Issues and/or People
- JIRA ticket(s) - STANFORD-419

yvonnetangsu commented on packages/card/CtaCard.jsx:
```diff
@@ -0,0 +1,109 @@
+        hasLink ? "su-rs-pb-4" : "su-rs-pb-5"
+      )}
+    >
+      {eyebrow && (
+        <span
+          aria-hidden
+          className="su-type-1 su-text-black-60 su-font-semibold su-rs-mb-1"
+        >
+          {eyebrow}
+        </span>
+      )}
```
This pattern was recommended by SODA to us on the ADAPT project - though sometimes opinions might differ from engineer to engineer. Basically, since the eyebrow is above the heading, they recommended aria-hidden it, and then stick it into the h2 (the card title) below as screen reader only text.

broleomo replied:
would you mind adding a code comment about this just so we remember why we did the pattern in the future?

yvonnetangsu replied:
Of course - happy to!

yvonnetangsu commented on packages/card/CtaCard.jsx:
```diff
@@ -0,0 +1,109 @@
+            {title}
+          </a>
+        ) : (
+          <>
+            {eyebrow && <span className="su-sr-only">{`${eyebrow}:`}</span>}
+            {title}
+          </>
```
As mentioned above, this is where I stick the eyebrow into the h2 as invisible screen reader only text. So here it would read to the screen reader like eg "Learn: See how aid works".

yvonnetangsu commented on packages/card/CtaCard.jsx:
```diff
@@ -0,0 +1,109 @@
+            )}
+          >
+            {isRealExternalLink ? (
+              <ExternalArrowUnstyled
+                title="link is external"
```
If the link is external, I added a title for the SVG "link is external". If the link is internal, I aria-hidden it since the chevron icon is decorative only.

yvonnetangsu commented on packages/card/CtaCard.jsx:
```diff
@@ -0,0 +1,109 @@
+          {eyebrow}
+        </span>
+      )}
+      <h2 className="su-type-5 md:su-type-4 2xl:su-type-3 su-mb-0 su-font-sans su-text-black">
+        {hasLink ? (
+          // eslint-disable-next-line react/jsx-no-target-blank
+          <a
+            href={internalUrl || externalUrl}
+            target={isNewWindow ? "_blank" : undefined}
+            rel={isRealExternalLink ? "noopener nofollow" : undefined}
+            className="group-hocus-within:su-underline su-stretched-link su-text-black group-hocus-within:su-text-digital-red"
+          >
+            {eyebrow && <span className="su-sr-only">{`${eyebrow}:`}</span>}
+            {title}
+          </a>
```
I wrap the h2 around the link, then use the su-stretched-link class so the clickable area covers the whole card - this is also a pattern recommended by SODA.

broleomo reviewed (commented):
Left a couple tiny comments and then we discussed on Slack the dark mode versions of the cards that will need some style updates so the backgrounds of the cards are black on dark mode, gradient buttons change to the green gradient, hocus color changes to the digital light red.  Here's the design file as well where I mocked it up if it's helpful. Thanks Yvonne great work! https://www.figma.com/file/1L5yoJvkNp2RSpX9RCwuI2/Designs----Campaign?type=design&node-id=2731-5709&mode=design&t=au00lYg8mG9ND64p-0

yvonnetangsu commented on packages/card/CtaCard.jsx:
```diff
@@ -0,0 +1,109 @@
+  isNewWindow,
+}) {
+  const hasLink = !!internalUrl || !!externalUrl;
+  // Check extenalUrl field to see if it is actually external
```
Oops - thanks 😹

yvonnetangsu commented:
@broleomo @iamrentman This is ready for your eyes again. I followed the dark mode mock and gave it a 2px thick border - I think it looks great with the dark mode, but on light mode, 1px border looks much more crisp - hopefully that isn't an issue. I've corrected the typos and added documentation about the eyebrow treatment as well. LMK if you see anything else 😄

yvonnetangsu commented on packages/card/CtaCard.jsx:
```diff
@@ -0,0 +1,115 @@
+        hasLink ? "su-rs-pb-4" : "su-rs-pb-5"
+      )}
+    >
+      {eyebrow && (
+        /**
+         * This is a SODA recommended pattern for accessibility.
+         * When there is an eyebrow, we aria-hidden it from screen readers,
+         * and then add the eyebrow text as a visually hidden span for screen readers inside the h2 heading below.
+         * This way we don't get an orphaned span that appears before the heading in the card.
+         */
+        <span
+          aria-hidden
+          className="su-type-1 su-text-black-60 su-font-semibold su-rs-mb-1"
+        >
+          {eyebrow}
+        </span>
```
Added documentation about this pattern here :)

broleomo commented:
Hey Yvonne would you be able to update the color of the icons on the gradient buttons in dark mode to be black instead of white? Outside of that, this PR looks great and is ready to merge in. 

cc: @JBCSU @iamrentman

iamrentman reviewed (approved):
Looks great. Thank you Yvonne!

yvonnetangsu commented:
Thank you both for the review! 😸
